README.md
# panforte

Отладка XSL шаблонов с выводом последовательности применения шаблонов,
контекста к которому они применялись и профайлинга по времени.

В данный момент proof-of-concept.

Для работы потребуется lxml.

## Использование

`python lxml_debug.py XSL XML`

Или с exslt расширениями, аналогичным используемыми в HeadHunter:

`python lxml_debug.py XSL XML --hh-exslt`

Проверено на:
* python=2.6 + lxml=2.2.8
* python=2.7 + lxml=3.2.4
* python=2.7 + lxml=2.2.8

## TODO

См. [список в lxml_debug.py](lxml_debug.py).

## Лицензция

The MIT License


## Демо

```bash
git clone https://github.com/maizy/panforte.git
cd panforte/
python lxml_debug.py samples/simple-xsl-xslt1/pages/breakfast.xsl samples/data/food-menu.xml
```


Выдаёт такой отчёт:
```
======== XSL FILE ========
samples/simple-xsl-xslt1/pages/breakfast.xsl
======== DATA FILE ========
samples/data/food-menu.xml
======== preprocessed root xsl ========
> processing /Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/item.xsl
> processing /Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/description.xsl
> processing /Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/description-tag.xsl
> processing /Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/calories.xsl
> reuse /Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/description-tag.xsl
======== RESULTS ========
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xmlns:pn="https://github.com/maizy/panforte">
  <body>
    <h1>Breakfast</h1>
    <h3>Belgian Waffles</h3>
    <span style="color: gray;">$5.95</span>
    <br/>
    <div class="description">Two of our famous Belgian Waffles with plenty of real maple syrup</div>
    <br/>
    <span style="color: orange;">650</span>
    <br/>
    <h3>Strawberry Belgian Waffles</h3>
    <span style="color: gray;">$7.95</span>
    <br/>
    <div class="description">Light Belgian waffles covered with strawberries and whipped cream</div>
    <br/>
    <span style="color: red;">900</span>
    <br/>
    <h3>Русский текст</h3>
    <span style="color: gray;">$8.95</span>
    <br/>
    <div class="description">Light Belgian waffles covered with an assortment of fresh berries and whipped cream</div>
    <br/>
    <span style="color: red;">900</span>
    <br/>
    <h3>French Toast</h3>
    <span style="color: gray;">$4.50</span>
    <br/>
    <div class="description">Thick slices made from our homemade sourdough bread</div>
    <br/>
    <h3>Homestyle Breakfast</h3>
    <span style="color: gray;">$6.95</span>
    <br/>
    <div class="description">Two eggs, bacon or sausage, toast, and our ever-popular hash browns</div>
    <br/>
  </body>
</html>

======== PROFILE ========
<profile>
  <template rank="1" match="food" name="" mode="food_menu" calls="5" time="21" average="4"/>
  <template rank="2" match="*" name="" mode="description-tag" calls="5" time="10" average="2"/>
  <template rank="3" match="" name="price-label" mode="" calls="5" time="9" average="1"/>
  <template rank="4" match="food" name="" mode="description" calls="5" time="8" average="1"/>
  <template rank="5" match="" name="calories-label" mode="" calls="3" time="8" average="2"/>
  <template rank="6" match="/breakfastMenu" name="" mode="" calls="1" time="8" average="8"/>
  <template rank="7" match="food[key('calories', @key) &gt;= 800]" name="" mode="calories" calls="2" time="4" average="2"/>
  <template rank="8" match="food" name="" mode="calories" calls="2" time="4" average="2"/>
  <template rank="9" match="*" name="" mode="path" calls="1" time="4" average="4"/>
  <template rank="10" match="food[key('calories', @key) &gt;= 650]" name="" mode="calories" calls="1" time="2" average="2"/>
</profile>
None
======== MESSAGES ========
TPL: context="/breakfastMenu", mode="None", match="/breakfastMenu", file="samples/simple-xsl-xslt1/pages/breakfast.xsl:12"
TPL: context="/breakfastMenu", mode="path", match="*", file="samples/simple-xsl-xslt1/pages/breakfast.xsl:24"
MSG: mes="/breakfastMenu", column="0", file="<string>:0"
TPL: context="/breakfastMenu/food[1]", mode="food_menu", match="food", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/item.xsl:6"
NAMED TPL: context="/breakfastMenu/food[1]", name="price-label", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/item.xsl:13"
TPL: context="/breakfastMenu/food[1]", mode="description", match="food", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/description.xsl:8"
TPL: context="/breakfastMenu/food[1]/description", mode="description-tag", match="*", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/description-tag.xsl:6"
TPL: context="/breakfastMenu/food[1]", mode="calories", match="food[key('calories', @key) >= 650]", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/calories.xsl:8"
MSG: mes="bw650", column="0", file="<string>:0"
NAMED TPL: context="/breakfastMenu/food[1]", name="calories-label", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/calories.xsl:22"
TPL: context="/breakfastMenu/food[2]", mode="food_menu", match="food", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/item.xsl:6"
NAMED TPL: context="/breakfastMenu/food[2]", name="price-label", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/item.xsl:13"
TPL: context="/breakfastMenu/food[2]", mode="description", match="food", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/description.xsl:8"
TPL: context="/breakfastMenu/food[2]/description", mode="description-tag", match="*", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/description-tag.xsl:6"
TPL: context="/breakfastMenu/food[2]", mode="calories", match="food[key('calories', @key) >= 800]", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/calories.xsl:15"
MSG: mes="850", column="0", file="<string>:0"
NAMED TPL: context="/breakfastMenu/food[2]", name="calories-label", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/calories.xsl:22"
TPL: context="/breakfastMenu/food[3]", mode="food_menu", match="food", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/item.xsl:6"
NAMED TPL: context="/breakfastMenu/food[3]", name="price-label", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/item.xsl:13"
TPL: context="/breakfastMenu/food[3]", mode="description", match="food", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/description.xsl:8"
TPL: context="/breakfastMenu/food[3]/description", mode="description-tag", match="*", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/description-tag.xsl:6"
TPL: context="/breakfastMenu/food[3]", mode="calories", match="food[key('calories', @key) >= 800]", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/calories.xsl:15"
MSG: mes="850", column="0", file="<string>:0"
NAMED TPL: context="/breakfastMenu/food[3]", name="calories-label", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/calories.xsl:22"
TPL: context="/breakfastMenu/food[4]", mode="food_menu", match="food", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/item.xsl:6"
NAMED TPL: context="/breakfastMenu/food[4]", name="price-label", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/item.xsl:13"
TPL: context="/breakfastMenu/food[4]", mode="description", match="food", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/description.xsl:8"
TPL: context="/breakfastMenu/food[4]/description", mode="description-tag", match="*", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/description-tag.xsl:6"
TPL: context="/breakfastMenu/food[4]", mode="calories", match="food", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/calories.xsl:30"
MSG: mes="default", column="0", file="<string>:0"
TPL: context="/breakfastMenu/food[5]", mode="food_menu", match="food", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/item.xsl:6"
NAMED TPL: context="/breakfastMenu/food[5]", name="price-label", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/item.xsl:13"
TPL: context="/breakfastMenu/food[5]", mode="description", match="food", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/description.xsl:8"
TPL: context="/breakfastMenu/food[5]/description", mode="description-tag", match="*", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/description-tag.xsl:6"
TPL: context="/breakfastMenu/food[5]", mode="calories", match="food", file="/Users/nikita/Dev/panforte/samples/simple-xsl-xslt1/blocks/calories.xsl:30"
MSG: mes="default", column="0", file="<string>:0"
```


CHANGELOG.md
# Changelog

# 0.0.2-alpha (2013-11-28)

- add hh-exslt extinsions


# 0.0.1-alpha (2013-11-28)

- proof-of-concept realization
- xml & simple xslt examples
- add hh-exslt extinsions, more debug in


lxml_apply.py
#!/usr/bin/env python
# _*_ coding: utf-8 _*_
# Copyright (c) Nikita Kovaliov, maizy.ru, 2013
# See LICENSE.txt for details.

from __future__ import unicode_literals, absolute_import, print_function

import sys
import os

from lxml import etree


def _title(text):
    print('{0} {1} {0}'.format('=' * 8, text))


def apply_xsl(xsl_file, data_file):
    _title('XSL FILE')
    print(xsl_file)

    _title('DATA FILE')
    print(data_file)

    data = etree.parse(open(data_file, 'r'))
    xsl = etree.parse(open(xsl_file, 'r'))

    transform = etree.XSLT(xsl)
    try:
        result = transform(data)
    except etree.XSLTApplyError as e:
        _title('XSLT Error')
        print(e)
        print('{e.filename}:{e.line}'.format(e=e.error_log.last_error))
        return 1

    _title('RESULTS')
    print(result)

    _title('MESSAGES')
    for entry in transform.error_log:
        print('message from line {en.line}, col {en.column}, file {en.filename}: {en.message}'
              .format(en=entry))
        print()


def main(argv):
    if len(argv) < 3:  # TODO: opt parse
        sys.stderr.write('Usage {}: xsl_file data_file\n'.format(argv[0]))
        return 2
    xsl_file = argv[1]
    data_file = argv[2]
    assert os.path.isfile(xsl_file)
    assert os.path.isfile(data_file)
    apply_xsl(xsl_file, data_file)
    return 0


if __name__ == '__main__':
    sys.exit(main(sys.argv))


lxml_debug.py
#!/usr/bin/env python
# _*_ coding: utf-8 _*_
# Copyright (c) Nikita Kovaliov, maizy.ru, 2013
# The MIT License, see LICENSE.txt for details.

from __future__ import unicode_literals, absolute_import, print_function

import sys
import os
import random
import cPickle as pickle
import base64
import pprint

from lxml import etree

__version__ = '0.0-proof-of-concept'

# TODOs:
#  * эмулировать файловые объекты, для вывода ошибок в XSLT с файлом и строкой
#  * opt parse
#  * output dumping ("-o")
#  * параметр "--base-path" для resolver, запрещающий выход за её пределы
#  * параметр "--dump-results" для сохранения в output обработанных xsl (только с base-path)
#  * logging
#  * api: возврат всех результатов в виде объектов (list'ы из namedtuple)
#    * xml вывод из результатов
#    * html отчёт из результатов
#  * api: хуки для пре/пост процессинга xsl, data xml, результатов
#  * подмешивать результаты стд. xslt профайлера
#  * unit тесты
#  * python 2.6, 2.7, 3.2, 3.3 support
#  * lxml 2.x & 3.x support
#  * оптимизация процессинга по времени и расходу памяти

dump_output = False  # XXX: не включать! не безопасно для примеров не из /samples


def main(argv):
    if len(argv) < 3:  # TODO: opt parse
        sys.stderr.write('Usage {0}: xsl_file data_file\n [--hh-exslt]'.format(argv[0]))
        return 2
    xsl_file = argv[1]
    data_file = argv[2]
    if len(argv) > 3 and argv[3] == '--hh-exslt':
        hh_exslt = True
    else:
        hh_exslt = False
    assert os.path.isfile(xsl_file)
    assert os.path.isfile(data_file)
    return debug(xsl_file, data_file, hh_exslt)


def _title(text):
    print('{0} {1} {0}'.format('=' * 8, text))


def _log(text):
    print('> ' + text)


def _strip_long_string(s, max_width=300):
    return s[:max_width-3] + '...' if len(s) > max_width else s

XSL_NS = 'http://www.w3.org/1999/XSL/Transform'
NSMAP = {'xsl': XSL_NS}
XSL_TAG = '{' + XSL_NS + '}'

PANFORTE_NS = 'https://github.com/maizy/panforte'
PANFORTE_TAG = '{' + PANFORTE_NS + '}'
PANFORTE_ALIAS = 'pn'
PANFORTE_NSMAP = {PANFORTE_ALIAS: PANFORTE_NS}

# FIXME: tmp
output_strip = os.path.abspath(os.path.dirname(__file__))
output_files = os.path.join(os.path.dirname(__file__), 'output')


def encode(debug_data):
    return base64.encodestring(pickle.dumps(debug_data))


def decode(encoded_debug_data):
    return pickle.loads(base64.decodestring(encoded_debug_data))


def add_xsl_debug_info(tree, file_name, prefix):
    for t_xml in tree.getroot().xpath('//xsl:template', namespaces=NSMAP):
        dbg_xml = etree.Element('{0}message'.format(XSL_TAG))
        debug_data = {
            b'l': t_xml.sourceline,
            b'f': file_name,
            b'n': t_xml.get('name'),
            b'm': t_xml.get('mode'),
            b'mt': t_xml.get('match'),
        }
        dbg_xml.text = '{0}|{1}|'.format(prefix, encode(debug_data))
        etree.SubElement(dbg_xml, '{0}value-of'.format(XSL_TAG), select='pn:path(current())')

        after = None
        for ch in t_xml.iterchildren():  # TODO: optimize
            if ch.tag in ('{0}param'.format(XSL_TAG), '{0}variable'.format(XSL_TAG), etree.Comment):
                after = ch
        if after is not None:
            after.addnext(dbg_xml)
        else:
            t_xml.insert(0, dbg_xml)


def dump_xsl(url, content):
    rel = os.path.relpath(url, output_strip)
    res_path = os.path.join(output_files, rel)
    try:
        os.makedirs(os.path.dirname(res_path))
    except OSError:
        pass
    _log('write xsl dump to {0}'.format(res_path))
    with open(res_path, 'w') as f:
        f.write(content)


class PanforteResolver(etree.Resolver):

    def __init__(self, prefix):
        super(PanforteResolver, self).__init__()
        self._prefix = prefix
        self.clean_up()

    def clean_up(self):
        self._processed = {}

    def resolve(self, url, pubid, context):
        if url not in self._processed:
            _log('processing {0}'.format(url))
            with open(url, 'r') as f:
                tree = etree.parse(f)
            add_xsl_debug_info(tree, url, self._prefix)
            res = etree.tostring(tree, encoding=unicode).encode('utf-8')
            res = add_panforte_xmlns(res)  # FIXME: hack inside

            # FIXME: tmp
            if dump_output:
                dump_xsl(url, res)

            self._processed[url] = res
        else:
            _log('reuse {0}'.format(url))
        return self.resolve_string(string=self._processed[url], context=context)


def add_panforte_xmlns(result_string):
    #FIXME hack
    return result_string.replace(b'<xsl:stylesheet',
                                 b'<xsl:stylesheet xmlns:{0}="{1}" '.format(PANFORTE_ALIAS, PANFORTE_NS))


def print_path(context, element, **kwargs):
    el = element[0]  # FIXME: why??
    if isinstance(el, basestring):  # text nodes
        parent = el.getparent()
        return 'TEXT OF: ' + el.getparent().getroottree().getpath(parent)
    else:
        return el.getroottree().getpath(el)


def debug(xsl_file, data_file, hh_exslt=False):

    _title('XSL FILE')
    print(xsl_file)

    _title('DATA FILE')
    print(data_file)

    prefix = b'panforte{0}'.format(random.randint(10000, 99999))
    xsl = etree.parse(open(xsl_file, 'r'))

    add_xsl_debug_info(xsl, xsl_file, prefix)
    xsl_str = etree.tostring(xsl, encoding=unicode).encode('utf-8')
    xsl_str = add_panforte_xmlns(xsl_str)  # FIXME: hack inside

    _title('preprocessed root xsl')
    if dump_output:
        dump_xsl(os.path.abspath(xsl_file), xsl_str)

    parser = etree.XMLParser()
    parser.resolvers.add(PanforteResolver(prefix))
    # TODO: don't do tostring/fromstring mangling
    xsl_tree = etree.fromstring(xsl_str, parser=parser, base_url=os.path.abspath(xsl_file))

    data = etree.parse(open(data_file, 'r'))

    ns = etree.FunctionNamespace(PANFORTE_NS)
    ns.prefix = PANFORTE_ALIAS
    ns['path'] = print_path

    if hh_exslt:
        import hh_json
        hh_json.etree_enrich_hh_namespace()

    etree.XPathEvaluator(data, namespaces=PANFORTE_NSMAP)

    transform = etree.XSLT(xsl_tree)
    try:
        result = transform(data, profile_run=True)
    except etree.XSLTApplyError as e:
        _title('XSLT Error')
        print(e)
        print(e.__dict__)
        return 1

    _title('RESULTS')
    print(result)

    _title('PROFILE')
    print(etree.dump(result.xslt_profile.getroot()))

    _title('MESSAGES')
    for entry in transform.error_log:
        mes = entry.message
        if mes.startswith(prefix):
            parts = mes.split('|', 3)
            debug_info = decode(parts[1])
            if debug_info['m'] is None and debug_info['mt'] is None:
                tpl = 'NAMED TPL: context="{c}", name="{n}", file="{f}:{l}"'
            else:
                tpl = 'TPL: context="{c}", mode="{m}", match="{mt}", file="{f}:{l}"'
            print(tpl.format(c=parts[2], **dict((k, _strip_long_string(v) if isinstance(v, basestring) else v)
                                                for k, v in debug_info.iteritems())))
        else:
            print('MSG: mes="{en.message}", column="{en.column}", file="{en.filename}:{en.line}"'
                  .format(en=entry))

    return 0

if __name__ == '__main__':
    sys.exit(main(sys.argv))
